OpenAPI spec for official registry API

tadasant opened the pull request:
This is a rough first pass at what the API shape should look like. Not meant to be final, we may not even merge this particular PR; just meant to give readers a sense of what we're going for to spark feedback conversations.

olaservo commented on api/README.md:
```diff
@@ -0,0 +1,359 @@
+  "id": "a5e8a7f0-d4e4-4a1d-b12f-2896a23fd4f1",
+  "name": "@modelcontextprotocol/servers/src/filesystem",
+  "description": "Node.js server implementing Model Context Protocol (MCP) for filesystem operations.",
+  "version": "0.0.3",
```
Is this `version` intended to match the package version when applicable, and should it be verified against the server version?

Bringing this up here related to [this](https://github.com/orgs/modelcontextprotocol/discussions/340).

tadasant replied:
We were actually planning to have a separate notion of version. Thoughts here: https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2074198582

Curious to hear if you think we should reconsider

tadasant replied:
ea6623e1a8a6efe87fb2e879ef953997b0f2016e

annaji-msft commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+      required:
+        - transport_type
+        - url
+      properties:
```
An optional, OAuth server configuration could help here as well specifically for those MCP authorizations servers that do not support Dynamic Client Registration. 

Described in the [spec](https://modelcontextprotocol.io/specification/draft/basic/authorization#2-5-dynamic-client-registration ), Clients could hardcode the ClientId for the authorization server. 

The metadata in the registry can help guide consumers a ClientId is expected for the specific MCP server as well as where to acquire it, usually service provider developer portals.

tadasant replied:
I'm thinking it might be helpful to lift the Remote shape into a separate PR/discussion. This Remote shape has not had much group input, and I haven't invested too much time in thinking through it + it was drafted a couple months ago before remotes grew in popularity and support. Agree that thinking through what auth-related concerns are worth including here is a good idea.

annaji-msft replied:
Thanks @tadasant happy to participate in the next iteration.

annaji-msft replied:
for tracking https://github.com/modelcontextprotocol/registry/issues/25

tadasant commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+          type: string
+          example: "info"
+    
+    Package:
+      type: object
+      required:
+        - registry_name
+        - name
+      properties:
+        registry_name:
+          type: string
+          enum: [npm, docker, pypi, homebrew]
+          example: "npm"
```
Ah I see, sorry I misunderstood your suggestion; I thought you were suggesting submission of a shell script separate from this shape. I see per your example that we'd still well-define the script within `command`.

> assuming there is a package-level version, the "NPM package name" should not get blindly added to the command because the version needs to be mixed in there as well. 

That's a good point. It should certainly be at least best practice to encode a version inside `command.positional_arguments.default_value`, like you did in your second example, `@modelcontextprotocol/server-filesystem@1.2.3`. Would be even better to enforce that somehow, so people aren't creating that implicit `@latest` inconsistency across registry version vs. MCP version... Need to think more on that.

> In our past thinking for the VS Code registry, arguments only specify the arguments for the package itself and not npx/uvx/whatever runner the client uses. Instead, it's up to the client to use a tool of its choice to retrieve the package at the specific version from the registry. This is also good because in cases where there is not a canonical way to install and run packages in that language (e.g. in Python, uvx is just one way to do things, and dotnet does not yet have any official such tool) the client is able to feature detect what capabilities are on the machine and do the right thing.

The thinking behind being explicit with `command.name` was that it limits the surface area for malicious commands and spam (there is only so much you can do with an enum type that has to be one of a few dozen options, like `npm`, `uvx`, etc.).

That said, a prior iteration of this shape did entirely omit `command.name`, in an attempt to do as you said:

> arguments only specify the arguments for the package itself and not npx/uvx/whatever runner the client uses

An example use case I found troubling was the case where some MCP servers will fail to run if you just do:

```
npx org@myserver --stdio
```

But will succeed if you do:

```
npx -y org@myserver --stdio
```

`-y` is an `npx` argument here.

But I don't feel strongly on this. Your points here make me think we should go back to omitting `name` and just let client apps figure out the `-y` nuances if they decide to use `npx` to run npm packages.

@alexhancock I think it was you and I that had the discussion that led to the addition of `name`. What do you think?

Found that discussion here: https://github.com/modelcontextprotocol/registry/pull/3/files#r2018897939

I am open to rethinking this. Maybe we could solve both this and the prior version-related concern by doing the following:
- Remove `command.name`
- Instead of putting the package name as the 0th positional argument, omit that
- Introduce a `version` field (required?)
- That means the client can create its preferred cocktail of `npx` or `uvx` or `pip`, whatever, knowing the package name and version should be enough to construct the first part of the command in a repeatable way
- Then the rest of the MCP-specific args can be constructed from this API shape (like Connor said, "arguments only specify the arguments for the package itself")

I think that approach should address your reason for wanting `shell`:

> I think they cannot both do this and properly respect the package-level verson, assuming one gets added.

So a client could see, for example:

```
      "registry_name": "npm",
      "name": "@modelcontextprotocol/server-filesystem",
      "version": "1.2.3",
      "command": {
        "subcommands": [],
        "positional_arguments": [
          {
            "position": 0,
            "name": "root_path",
            "description": "Root path for filesystem access",
            "default_value": "/projects",
            "is_required": true,
            "is_editable": true,
            "is_repeatable": false,
            "choices": []
          }
```

And then construct, based on its preference for `npx`:

```sh
npx -y @modelcontextprotocol/server-filesystem@1.2.3 /projects
```

connor4312 replied:
> So a client could see, for example:

Yep, exactly. Your summary accurately reflects my thinking and suggestions on this 🙂 

> Introduce a version field (required?)

Yes, I think this should should be required (as per my other comments)

sandy081 replied:
```json
      "registry_name": "npm",
      "name": "@modelcontextprotocol/server-filesystem",
      "version": "1.2.3",
      // ... args
```

+1 to the above syntax and this aligns with our past thinking in VS Code as @connor4312 mentioned. 

Since we are delegating the command to run to the client, how about simplifying the command/subcommands/positional_arguments/named_arguments to just `arguments` array as a property under `package` object?  For example an argument object schema can be as follows:

```json
{
			"type": "object",
			"description": "Describes an argument that can be passed to the server",
			"required": [
				"name"
			],
			"properties": {
				"name": {
					"type": "string",
					"description": "Name of the argument"
				},
				"value": {
					"oneOf": [
						{
							"type": "string",
							"description": "Value of the argument"
						},
						{
							"type": "array",
							"description": "List of values for the argument",
							"items": {
								"type": "string"
							}
						}
					]
				},
				"description": {
					"type": "string",
					"description": "Description of the argument"
				},
				"default": {
					"type": ["string", "array"],
					"description": "Default value of the argument"
				},
				"is_required": {
					"type": "boolean",
					"description": "If true, the argument is required"
				},
                                 "is_value_required": {
					"type": "boolean",
					"description": "If true, the value is required for the argument"
				},
				"choices": {
					"type": "array",
					"description": "List of possible values for the argument",
					"items": {
						"type": "string",
						"description": "Possible value for the argument"
					}
				}
			}
		}
```

A sample manifest can be as follows:

```json
      "registry_name": "npm",
      "name": "@modelcontextprotocol/server-filesystem",
      "version": "1.2.3",
     "arguments": [
          {
            "name": "root_path",
            "description": "Root path for filesystem access",
            "default": "/projects",
            "is_required": true,
            "is_value_required": true,
          }
      ]        
```

This will let the client just to pass the arguments passed in the manifest as is to the server (including headers, env variables) by validating some flags like `is_required`. 

What do you think?

alexhancock replied:
@connor4312 @tadasant Late to the party on this thread but the conclusion you got to in https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2077835152 makes sense to me and I agree

On the topic of flattening into one list of args, it could simplify things. I think it could still work for positional as well as named arguments. For example we use a single arg list in our current approach to configuring the client to run MCP servers in goose. Example YAML:

```yaml
figma:
  args:
    - -y
    - 'imaginary-figma-mcp@latest'
  bundled: true
  cmd: npx
  ... (rest omitted)
```

But I may be missing something. What issue do you see @tadasant ?

connor4312 replied:
The actual client config of servers is out of scope of the registry spec, but my plan for VS Code is to have something like (on a data level)

```
{
  "server": "imaginary-figma-mcp@1.2.3",
  "type": "npm", // optionally diambiguate
  "config": { "apiKey": "something-here" }
}
```

...and generate the CLI behind the scenes such that versions can be updated and largely tolerant to argument changes so long as inputs are stable.

tadasant replied:
I'm having a doubt on removing `subcommand`: how would we then make the [filesystem reference server work with Docker](https://github.com/modelcontextprotocol/servers/tree/HEAD/src/filesystem#docker)?

```json
{
  "mcpServers": {
    "filesystem": {
      "command": "docker",
      "args": [
        "run",
        "-i",
        "--rm",
        "--mount", "type=bind,src=/Users/username/Desktop,dst=/projects/Desktop",
        "--mount", "type=bind,src=/path/to/other/allowed/dir,dst=/projects/other/allowed/dir,ro",
        "--mount", "type=bind,src=/path/to/file.txt,dst=/projects/path/to/file.txt",
        "mcp/filesystem",
        "/projects"
      ]
    }
  }
}
```

Maybe this example is a major outlier and we shouldn't design around it, but I wouldn't feel great about not even supporting our official reference servers.

What if we tried to simplify using a templating solution? e.g. something like:

```
{
  "id": "filesystem-67890",
  "name": "@modelcontextprotocol/servers/src/filesystem",
  "description": "Node.js server implementing Model Context Protocol (MCP) for filesystem operations",
  "repository": { "url": "https://github.com/modelcontextprotocol/servers", "source": "github", "id": "d94b5f7e-c7c6-d760-2c78-a5e9b8a5b8c9" },
  "version_detail": { "version": "1.0.2", "release_date": "2023-06-15T10:30:00Z", "is_latest": true },
  "package_canonical": "docker",
  "packages": [
    {
      "registry_name": "docker",
      "name": "mcp/filesystem",
      "command_templates": {
        "docker": "docker run -i --rm --mount {{MOUNT_CONFIG}} {{IMAGE_NAME}} {{ROOT_PATH}}",
        "podman": "podman run -i --rm --mount {{MOUNT_CONFIG}} {{IMAGE_NAME}} {{ROOT_PATH}}"
      },
      "preferred_command": "docker",
      "variables": [
        {
          "name": "MOUNT_CONFIG",
          "description": "Mount a volume into the container",
          "default_value": "type=bind,src=/Users/username/Desktop,dst=/projects/Desktop",
          "is_required": true,
          "is_editable": true,
          "is_repeatable": true
        },
        {
          "name": "IMAGE_NAME",
          "description": "Docker image name",
          "default_value": "mcp/filesystem",
          "is_required": true,
          "is_editable": false
        },
        {
          "name": "ROOT_PATH",
          "description": "Root path for filesystem access",
          "default_value": "/projects",
          "is_required": true,
          "is_editable": false
        }
      ],
      "environment_variables": [
        {
          "name": "LOG_LEVEL",
          "description": "Logging level (debug, info, warn, error)",
          "required": false,
          "default_value": "info"
        }
      ]
    }
  ]
}
```

tadasant replied:
https://github.com/modelcontextprotocol/registry/issues/28

tadasant commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+              type: string
+              enum: [npm, docker, pypi, crates]
+              example: "npm"
+            packages:
```
Yes, it is an optional field in this spec.

luoxiner commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+          $ref: '#/components/schemas/Repository'
+        version:
+          type: string
+          description: Dynamically pulled from registry_canonical reference
```
Is the "version" here related to the [#/definitions/Implementation](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/c255349bd98d9dc73f3579c828d195bd09c0bee8/schema/2025-03-26/schema.json#L586) version field?

Does the "version" field here convey the differences resulting from different implementations of the MCP server?
For instance, adding a tool, removing a tool, or modifying the functions of a tool. When the capabilities of the MCP server itself change, the version here needs to be adjusted accordingly. 

Does the specific version information of the package need to be directly included in the Package?
For the same version of the MCP server, there are different versions of package implementations, but their capabilities are the same.

tadasant replied:
> Is the "version" here related to the [#/definitions/Implementation](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/c255349bd98d9dc73f3579c828d195bd09c0bee8/schema/2025-03-26/schema.json#L586) version field?

Good callout - I think yes, it should match, but we haven't deeply considered the relationship.

> Does the "version" field here convey the differences resulting from different implementations of the MCP server?
> For instance, adding a tool, removing a tool, or modifying the functions of a tool. When the capabilities of the MCP server itself change, the version here needs to be adjusted accordingly.

Yes, that's the idea. The published "mcp.json" file should be immutable. So any change in that (including a source code change, which would drive the kinds of changes you listed) should be communicated with a version bump and new mcp.json publication.

> Does the specific version information of the package need to be directly included in the Package?
For the same version of the MCP server, there are different versions of package implementations, but their capabilities are the same.

I'm not sure I follow what you are asking here, mind rephrasing?

luoxiner replied:
I'm sorry for the confusion in my previous question. What I meant is that the version here should not depend on any specific package version, and the version of each package should be defined solely in the Package.version field.

luoxiner replied:
I'm sorry for the confusion in my previous question. What I meant is that the version here should not depend on any specific package version, and the version of each package should be defined solely in the Package.version field.

But I believe you already addressed the question in your previous reply.

tadasant replied:
Removed the errant language here ea6623e1a8a6efe87fb2e879ef953997b0f2016e

luoxiner commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+          schema:
+            type: string
+            format: uuid
+        - name: version
```
I reviewed the current Go implementation and noticed that a UUID is generated every time an MCP server is published, ensuring that each version of an MCP server has a unique ID. 

However, it seems that the id here is associated only with the MCP server itself, rather than with the server + version, which is causing some confusion for me. 

And I think the current specification is more reasonable for a clear semantic understanding.

tadasant replied:
The implementation is very much a work in progress, so it's probably just a bug, but cc @sridharavinash to confirm

sandy081 commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+        url:
+          type: string
+          format: uri
+          example: "https://mcp-fs.example.com/sse"
```
Yes, it is similar to local package arguments. This is how we envisioned for example:

```json
"remote": {
			"type": "object",
			"description": "Remote server configuration",
			"required": [
				"url"
			],
			"properties": {
				"url": {
					"type": "string",
					"format": "uri",
					"description": "URL of the remote server"
				},
				"transport_type": {
					"type": "string",
					"description": "Transport type for the remote server",
					"enum": [
						"streamable",
						"sse"
					]
				},
				"headers": {
					"type": "array",
					"description": "Headers to be sent to the remote server",
					"items": { "$ref": "#/definitions/argument" }
				}
			}
		}
```

`#/definitions/argument` is the schema for the argument that is also used for local package arguments.

sandy081 commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+        - id
+        - name
+        - description
+        - repository
```
I assume the `version` property will be made required here as per the below discussion.

tadasant replied:
e7f8916d504365e185e02ddb45a0807a58cd38a5

sandy081 commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+          type: string
+          format: uuid
+          example: "a5e8a7f0-d4e4-4a1d-b12f-2896a23fd4f1"
+        name:
```
Should we also have a `displayName` that can be readable to the user, say `File System MCP`?

tadasant replied:
To continue discussion: https://github.com/modelcontextprotocol/registry/discussions/27

sandy081 commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+        - name
+        - description
+        - repository
+      properties:
```
It should be good to include optional `icon` property in the first iteration so that mcp servers are nicely displayed in the client. Eg schema:

```
"icon": {
				"type": "object",
				"description": "Icon of the server",
				"required": [
					"url"
				],
				"properties": {
					"url": {
						"type": "string",
						"description": "URL of the icon"
					}
				}
			}
```

sandy081 replied:
Agreed that passing URL's is risky. To prevent this the MCP server registry should scan these URLs and filter them for security like not allowing SVGs and including icons while registering an MCP server. 

> should we allow for storing different sized icons for different displays?

Yeah we should going forward, but for the first cut an icon url to display in the gallery should be sufficient.

> Would it be reasonable for you to consider instead using, for example, the favicon of the reverse-DNS name? At least initially, and maybe we could add icon support natively later on when there are more eyeballs on this and we can iteratively improve it?

If this is the initial direction, then I would like the server to do this reverse engineering and provide the icon url in the server list response. Again, i am trying to prevent inconsistencies of representing an mcp server by letting server providing this information

connor4312 replied:
> To prevent this the MCP server registry should scan these URLs and filter them for security like not allowing SVGs and including icons while registering an MCP server.

This is not a trivial ask. It's also not really possible to do for external URLs, consider I might have `example.com/icon` which is a png when I submit the package but I change it to a an SVG later. Say that to fix that we spec that the URL must have an extension with a mimetype that matches the URL. I still control the URL so I could make `example.com/icon.png` redirect to `example.com/evil.svg`, which HTML-based clients in `<img` will follow by default. So we need client validation, or some really aggressive server behavior to continuously validate image. This is an very deep hole and there may be 3P servers as well which don't do any validation or have their own bugs.

Imo it should be on the client to validate the images it retrieves from the server and do the right thing.

For VS Code, we should probably run our own image proxy server anyway (like github camo) to avoid exposing user IP addresses to third parties and validation can be done there. But we should not require this of registry implementations.

sandy081 commented:
@tadasant / @sridharavinash 

May I know if the server definition supports a required `publisher` property? I remember @sridharavinash mentioning that this can be derived from the repository owner or org. If so then it makes sense that the server registry populates the publisher information in the server list and server response.

sandy081 commented:
>  There was no current plan to include a publisher property -- it is just implied through the reverse-DNS `name` field.

Is the `name` field meant for the name of the MCP server, eg: `@modelcontextprotocol/server-filesystem` in which case the owner/publisher would be `modelcontextprotocol` right? If so, it would be helpful if the server registry resolves this information into a separate property with necessary details like `publisher` and `publisherDisplayName`, for eg

```json
"publisher": {
    "name": "modelcontextprotocol",
    "displayName": "Model Context Protocol",
    "verified": true
}
```

I think the registry is the right one to resolve this information and leaving this to clients could cause falsy information. It is important for a user to know the publisher/owner of an MCP server. Hence it would be great if this can be included in the initial version.

tadasant commented on api/openapi.yaml:
```diff
@@ -0,0 +1,324 @@
+      properties:
+        transport_type:
+          type: string
+          enum: [streamable, sse, ws]
```
f9b0f3121a1021e7053e05bae9bd17f189568193

tadasant commented:
Thanks for all the feedback so far @alexhancock @connor4312 @sandy081 @luoxiner @annaji-msft 

I am thinking we have covered a good amount of breadth to start, and at this point there need to be some more detailed discussions about specific points, and it's feeling unwieldy to try to handle in a single monster PR here.

How would you all feel if we merged this OpenAPI spec and then spun up separate PR's/Issues to drill down on the open questions we've identified?

Specifically:
- Incorporating [OAuth details](https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2076841333) into the remote servers shape (@annaji-msft)
- Incorporating [headers for remotes](https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2074202871) (@connor4312)
- Consider including "marketing" data, like [displayName](https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2079321488) and [icon](https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2079330007). I had initially thought displayName might make sense, but I am now leaning towards thinking we should exclude both. Can share my thoughts in an Issue, would like to get more opinions on this one (@sandy081)
- Consider including [publisher data](https://github.com/modelcontextprotocol/registry/pull/3#issuecomment-2863341238). Can discuss more in an issue, but I think there a major issues here with the notion of "trust". We actually have no guarantees that the publisher is who they say they are. We've got the reverse DNS system planned, but all that is is the namespace. (@sandy081)
- Figuring out the [proper way to model commands](https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2080826697). I'm starting to think we should simplify into a templating approach. (@connor4312 @alexhancock @sandy081)

Besides these above open points, I've addressed feedback.

If you all think it's a good idea to merge this and start to separate out the discussions, I'll start an Issue for each of these, and if someone is eager to finalize they can open a PR without being blocked on me. Does that work?

cc @toby @digitarald

evalstate commented:
Thanks @tadasant  -- this is brilliant work btw. I was going to ask how to contribute, so I definitely +1 on getting this merged.

connor4312 commented:
> Figuring out the https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2080826697. I'm starting to think we should simplify into a templating approach.

I considered this in my initial draft of how a registry manifest would work. I thought of templating args, environment variables, and headers ('inputs') with Mustache-like syntax:

```js
"docker": {
  "configuration": {
    "required": ["github-token"],
    "github-token": {
      "type": "string",
      "description": "GitHub Personal Access Token",
      "secret": true
    }
  },
  "name": "ghcr.io/github/github-mcp-server",
  "args": ["-e", "GITHUB_TOKEN"],
  "env": { "GITHUB_TOKEN": "{{github-token}}" }
}
```

In my initial draft I stated that inputs that reference an optional configuration that's missing would be omitted. (Is this is enough? I'm not sure.) Then there are the usual niggles with templating, like how we handle escaping and so on. Configuration being a JSON schema object gives some more power but opens up to nested objects and arrays and I'm not sure how/if those should be handled (or if we'd spec it such that only primitives are allowed.) It also doesn't map cleanly onto repeated arguments unless we get deep into the weeds of templating.

Overall I liked the explicit version more. It kept the world better-bounded and gave clear answers to the above unknowns.  Sandeep [linked](https://github.com/modelcontextprotocol/registry/pull/3#discussion_r2079263686) roughly where we ended up on our version of the spec.

digitarald reviewed (approved):
LGTM based on https://github.com/modelcontextprotocol/registry/pull/3#issuecomment-2864970637

tadasant commented:
Follow-ups:
- https://github.com/modelcontextprotocol/registry/issues/26 - HTTP Headers. Pretty much shovel-ready if anyone wants to self-assign
- https://github.com/modelcontextprotocol/registry/issues/25 - Thanks @annaji-msft for opening this, I think it's close to shovel-ready
- https://github.com/modelcontextprotocol/registry/issues/28 - `command` shape. I think we should solicit some more opinions on this from folks before locking in a conclusion, but I think @connor4312 and @sandy081 have done a good amount of thinking here that it is probably close to being shovel-ready
- https://github.com/modelcontextprotocol/registry/issues/29 - Publisher metadata. I don't think we should proceed with publisher metadata, but feedback welcome
- https://github.com/modelcontextprotocol/registry/discussions/27 - icon & displayName. I don't think we should proceed with this, but feedback welcome

tiovikram commented on api/README.md:
```diff
@@ -0,0 +1,383 @@
+        ],
+        "positional_arguments": [
+          {
+            "position": 0,
```
Explicitly specifying the position as a key in the JSON seems like an anti-pattern given that the parent object (an array) is inherently ordered by nature.

What are your thoughts about this? Any chance I could open this up as a follow-up issue if you think it is pertinent?

connor4312 replied:
You may be interested in looking at this PR: https://github.com/modelcontextprotocol/registry/pull/33

tiovikram replied:
I did take a look and I think #33 may be simplifying the commands a little too much. However, I can move that discussion there instead.
